src/mesh/parallel_tree.jl

```diff
@@ -0,0 +1,833 @@
+
+  function ParallelTree{NDIMS}(capacity::Integer) where NDIMS
+    # Verify that NDIMS is an integer
+    @assert NDIMS isa Integer
+
+    # Create instance
+    t = new()
+
+    # Initialize fields with defaults
+    # Note: length as capacity + 1 is to use `capacity + 1` as temporary storage for swap operations
+    t.parent_ids = fill(typemin(Int), capacity + 1)
+    t.child_ids = fill(typemin(Int), 2^NDIMS, capacity + 1)
+    t.neighbor_ids = fill(typemin(Int), 2*NDIMS, capacity + 1)
+    t.levels = fill(typemin(Int), capacity + 1)
+    t.coordinates = fill(NaN, NDIMS, capacity + 1)
+    t.original_cell_ids = fill(typemin(Int), capacity + 1)
+    t.mpi_ranks = fill(typemin(Int), capacity + 1)
+
+    t.capacity = capacity
+    t.length = 0
+    t.dummy = capacity + 1
+
+    t.center_level_0 = @SVector fill(NaN, NDIMS)
+    t.length_level_0 = NaN
+
+    return t
+  end
+end
+
+
+# Constructor for passing the dimension as an argument
+ParallelTree(::Val{NDIMS}, args...) where NDIMS = ParallelTree{NDIMS}(args...)
+
+# Create and initialize tree
+function ParallelTree{NDIMS}(capacity::Int, center::AbstractArray{Float64},
+                 length::Real, periodicity=true) where NDIMS
+  # Create instance
+  t = ParallelTree{NDIMS}(capacity)
+
+  # Initialize root cell
+  init!(t, center, length, periodicity)
+
+  return t
+end
+
+# Constructor accepting a single number as center (as opposed to an array) for 1D
+ParallelTree{1}(cap::Int, center::Real, len::Real, periodicity=true) = ParallelTree{1}(cap, [convert(Float64, center)], len, periodicity)
+
+
+# Clear tree with deleting data structures, store center and length, and create root cell
+function init!(t::ParallelTree, center::AbstractArray{Float64}, length::Real, periodicity=true)
+  clear!(t)
+
+  # Set domain information
+  t.center_level_0 = center
+  t.length_level_0 = length
+
+  # Create root cell
+  t.length += 1
+  t.parent_ids[1] = 0
+  t.child_ids[:, 1] .= 0
+  t.levels[1] = 0
+  t.coordinates[:, 1] .= t.center_level_0
+  t.original_cell_ids[1] = 0
+  t.mpi_ranks[1] = typemin(Int)
+
+  # Set neighbor ids: for each periodic direction, the level-0 cell is its own neighbor
+  if all(periodicity)
+    # Also catches case where periodicity = true
+    t.neighbor_ids[:, 1] .= 1
+    t.periodicity = ntuple(x->true, ndims(t))
+  elseif !any(periodicity)
+    # Also catches case where periodicity = false
+    t.neighbor_ids[:, 1] .= 0
+    t.periodicity = ntuple(x->false, ndims(t))
+  else
+    # Default case if periodicity is an iterable
+    for dimension in 1:ndims(t)
+      if periodicity[dimension]
+        t.neighbor_ids[2 * dimension - 1, 1] = 1
+        t.neighbor_ids[2 * dimension - 0, 1] = 1
+      else
+        t.neighbor_ids[2 * dimension - 1, 1] = 0
+        t.neighbor_ids[2 * dimension - 0, 1] = 0
+      end
+    end
+
+    t.periodicity = Tuple(periodicity)
+  end
+end
+
+
+# Convenience output for debugging
+function Base.show(io::IO, t::ParallelTree{NDIMS}) where NDIMS
```

Comment:
I think it's more appropriate to make this a method

```suggestion
function Base.show(io::IO, ::MIME"text/plain", t::ParallelTree{NDIMS}) where NDIMS
```

As far as I know, plain `Base.show(io::IO, t::ParallelTree` should usually produce a single-line output.

Reply:
Thanks for the hint! Fixed in 1fa802f.